src/waltham/waltham-connection.h

```diff
@@ -373,6 +354,83 @@ wth_connection_get_protocol_error(struct wth_connection *conn,
 				  const char **interface,
 				  uint32_t *object_id);
 
+/** Return a registry using the internal wth_display
+ *
+ * \param conn The Waltham connection.
+ * \return The registry obtained.
+ *
+ * The returned registry can then be used to retrieve other global
+ * objects, using its usual interface.
+ *
+ * \memberof wth_connection
+ * \client_api
+ */
+struct wthp_registry *
+wth_connection_get_registry(struct wth_connection *conn);
+
+/** Return a callback using the internal wth_display
+ *
+ * \param conn The Waltham connection.
+ * \return The callback obtained.
+ *
+ * The returned callback will emit a 'done' event once the related
+ * request is done.
+ *
+ * \memberof wth_connection
+ * \client_api
+ */
+struct wthp_callback *
+wth_connection_sync(struct wth_connection *conn);
+
+/** Set the callback to be called once a client requests a registry
+ *
+ * \param conn The Waltham connection.
+ * \param callback The registry callback.
+ * \param user_pointer Any data passed by the client.
+ *
+ * The callback will be stored inside of the connection, and will be
+ * called each time a client requests a registry.
+ *
+ * \memberof wth_connection
+ * \server_api
+ */
+void
+wth_connection_set_registry_callback(struct wth_connection *conn,
+                                     void (*callback)(struct wthp_registry *, void *),
+                                     void *user_pointer);
+
+/** Send an error on a specific object
+ *
+ * \param conn The Waltham connection.
+ * \param obj A Waltham object.
+ * \param code The error code.
+ * \param str The error message.
+ *
+ * This function wraps the wth_display_send_error display function, so
+ * the server doesn’t have to use this private API.
+ *
+ * \memberof wth_connection
+ * \server_api
+ */
+void
+wth_connection_send_error(struct wth_connection *conn, struct wth_object *obj,
+                          uint32_t code, const char *str);
+
+/** Post a display error
+ *
+ * \param conn The Waltham connection.
+ * \param code The error code.
+ * \param fmt... The error message.
```

Comment:
Remove the `...`, it makes Doxygen grumpy. :-)
The documentation should emphasize that the error code must be one defined for `wth_display`.
Hm. That's a bit of a problem, as `wth_display` is now private protocol, implying that the API here is not quite right. Maybe we should have specific error functions instead, e.g. `wth_connection_post_error_no_memory(connection)`? As for invalid object and invalid method, I'm not even sure the users of libwaltham would need them strictly speaking. Protocol errors should be defined in the respective interfaces in XML, so there should be little need to send "generic" errors.